developer/src/kmcmplib/tests/gtest-compiler-test.cpp

```diff
@@ -0,0 +1,458 @@
+    KMX_WCHAR str[LINESIZE];
+
+    // CERR_NoTokensFound
+    str[0] = '\0';
+    EXPECT_EQ(CERR_NoTokensFound, ProcessBeginLine(&fileKeyboard, str));
+
+    // CERR_InvalidToken
+    u16cpy(str, u"abc >");
+    EXPECT_EQ(CERR_InvalidToken, ProcessBeginLine(&fileKeyboard, str));
+
+    // CERR_RepeatedBegin, BEGIN_UNICODE
+    kmcmp::BeginLine[BEGIN_UNICODE] = 0; // not -1
+    u16cpy(str, u" unicode>");
+    EXPECT_EQ(CERR_RepeatedBegin, ProcessBeginLine(&fileKeyboard, str));
+    kmcmp::BeginLine[BEGIN_UNICODE] = -1;
+
+    // CERR_RepeatedBegin, BEGIN_ANSI
+    kmcmp::BeginLine[BEGIN_ANSI] = 0; // not -1
+    u16cpy(str, u" ansi>");
+    EXPECT_EQ(CERR_RepeatedBegin, ProcessBeginLine(&fileKeyboard, str));
+    kmcmp::BeginLine[BEGIN_ANSI] = -1;
+
+    // CERR_RepeatedBegin, BEGIN_NEWCONTEXT
+    kmcmp::BeginLine[BEGIN_NEWCONTEXT] = 0; // not -1
+    u16cpy(str, u" newContext>");
+    EXPECT_EQ(CERR_RepeatedBegin, ProcessBeginLine(&fileKeyboard, str));
+    kmcmp::BeginLine[BEGIN_NEWCONTEXT] = -1;
+
+    // CERR_RepeatedBegin, BEGIN_POSTKEYSTROKE
+    kmcmp::BeginLine[BEGIN_POSTKEYSTROKE] = 0; // not -1
+    u16cpy(str, u" postKeystroke>");
+    EXPECT_EQ(CERR_RepeatedBegin, ProcessBeginLine(&fileKeyboard, str));
+    kmcmp::BeginLine[BEGIN_POSTKEYSTROKE] = -1;
+};
+
+TEST_F(CompilerTest, ValidateMatchNomatchOutput_test) {
+    EXPECT_EQ(CERR_None, ValidateMatchNomatchOutput(NULL));
+    EXPECT_EQ(CERR_None, ValidateMatchNomatchOutput((PKMX_WCHAR)u""));
+    const KMX_WCHAR context[] = { 'a', 'b', 'c', UC_SENTINEL, CODE_CONTEXT, 'd', 'e', 'f' };
+    EXPECT_EQ(CERR_ContextAndIndexInvalidInMatchNomatch, ValidateMatchNomatchOutput((PKMX_WCHAR)context));
+    const KMX_WCHAR contextex[] = { 'a', 'b', 'c', UC_SENTINEL, CODE_CONTEXTEX, 'd', 'e', 'f' };
+    EXPECT_EQ(CERR_ContextAndIndexInvalidInMatchNomatch, ValidateMatchNomatchOutput((PKMX_WCHAR)contextex));
+    const KMX_WCHAR index[] = { 'a', 'b', 'c', UC_SENTINEL, CODE_INDEX, 'd', 'e', 'f' };
+    EXPECT_EQ(CERR_ContextAndIndexInvalidInMatchNomatch, ValidateMatchNomatchOutput((PKMX_WCHAR)index));
+    const KMX_WCHAR sentinel[] = { 'a', 'b', 'c', UC_SENTINEL, 'd', 'e', 'f' };
+    EXPECT_EQ(CERR_None, ValidateMatchNomatchOutput((PKMX_WCHAR)sentinel));
+};
+
+// KMX_DWORD ParseLine(PFILE_KEYBOARD fk, PKMX_WCHAR str)
+// KMX_DWORD ProcessGroupLine(PFILE_KEYBOARD fk, PKMX_WCHAR p)
+// int kmcmp::cmpkeys(const void *key, const void *elem)
+// KMX_DWORD ProcessGroupFinish(PFILE_KEYBOARD fk)
+// KMX_DWORD ProcessStoreLine(PFILE_KEYBOARD fk, PKMX_WCHAR p)
+// bool resizeStoreArray(PFILE_KEYBOARD fk)
+// bool resizeKeyArray(PFILE_GROUP gp, int increment)
+// KMX_DWORD AddStore(PFILE_KEYBOARD fk, KMX_DWORD SystemID, const KMX_WCHAR * str, KMX_DWORD *dwStoreID)
+// KMX_DWORD AddDebugStore(PFILE_KEYBOARD fk, KMX_WCHAR const * str)
+// KMX_DWORD ProcessSystemStore(PFILE_KEYBOARD fk, KMX_DWORD SystemID, PFILE_STORE sp)
+// int GetCompileTargetsFromTargetsStore(const KMX_WCHAR* store)
```

Comment:
Yes ... all the functions need testing, and with so many in the Compiler.cpp file, I have put the signatures in here as comments so that as I add each set of tests they will be in the same order as the original functions.  These comments will thus gradually disappear over successive PRs.